src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -244,129 +259,136 @@ async def destination_pool_error(self, new_pool):
                     self.logger.warning("Failed to clear old core files.", exc_info=True)
 
-            await run('umount', '/var/lib/systemd/coredump', check=False)
+            subprocess.run(['umount', '/var/lib/systemd/coredump'], check=False)
             os.makedirs('/var/lib/systemd/coredump', exist_ok=True)
-            await run('mount', '--bind', corepath, '/var/lib/systemd/coredump')
+            subprocess.run(['mount', '--bind', corepath, '/var/lib/systemd/coredump'])
 
-        await self.middleware.call('etc.generate', 'glusterd')
+        self.middleware.call_sync('etc.generate', 'glusterd')
 
         if mounted:
-            # There is no need to wait this to finish
-            # Restarting rrdcached will ensure that we start/restart collectd as well
-            asyncio.ensure_future(self.middleware.call('service.restart', 'rrdcached'))
-            asyncio.ensure_future(self.middleware.call('service.restart', 'syslogd'))
+            self.middleware.call_sync('systemdataset._post_setup_service_restart')
 
-            await self.middleware.call('smb.setup_directories')
-            # The following should be backgrounded since they may be quite
-            # long-running.
-            await self.middleware.call('smb.configure', False)
+        return self.middleware.call_sync('systemdataset.config')
 
-        return await self.config()
+    def _query_pool_for_system_dataset(self, exclude_pool):
+        for p in self.middleware.call_sync('pool.query'):
+            if (exclude_pool and p['name'] == exclude_pool) or self.middleware.call_sync('pool.dataset.query', [
+                ['name', '=', p['name']], [
+                    'OR', [['key_format.value', '=', 'PASSPHRASE'], ['locked', '=', True]]
+                ]
+            ]):
+                continue
 
-    async def _query_pool_for_system_dataset(self, exclude_pool):
-        for name in await self._query_pools_names_for_system_dataset(exclude_pool):
-            return await self.middleware.call('pool.query', [['name', '=', name]], {'get': True})
+            return p
 
     async def _query_pools_names_for_system_dataset(self, exclude_pool=None):
         return [
-            ds['id']
-            for ds in await self.middleware.call(
-                'pool.dataset.query',
+            m['mount_source'].split('/')[0]
+            for m in await self.middleware.call(
+                'filesystem.mount_info',
                 [
+                    ['fs_type', '=', 'zfs'],
                     ['OR', [
-                        # Pools without encryption
-                        ['encrypted', '=', False],
-                        # Encrypted pools that are not locked
-                        ['locked', '=', False],
-                        # Passphrase-encrypted pools (system dataset will be unencrypted in this case)
-                        ['key_format.value', '=', 'PASSPHRASE'],
+                        ['mount_source', 'rnin', '/'],
```

Comment:
Previous code did not include key-encrypted pools, this code does.

Reply:
Previous did include mounted key-encrypted pools. I think in above case the only case that is now being excluded is unmounted passphrase-encrypted pools. I'm not sure if that's an intentional design feature, is it?

Reply:
We should be able to use a passphrase-encrypted pool for system dataset even if it is not unlocked yet.

Reply:
Okay. Fixing.